README.md
# ldm cookbook

* Installs ldm from source code
* Configures ldmd.conf, scour.conf, and pqact.conf based on attributes
* Adds cronjobs based on attributes

# Requirements

* chef
* CentOS 6 or Centos 7

# Usage

* to stop the ldm from being started set `node['ldm']['auto-start'] = false`

# Attributes
sample attribute from node json:
```
    "override_attributes": {
      "ldm": {
        "requests": [
  
        ],
        "allows": [
          {
            "feedset": "ANY",
            "hostname": "(10.10.10.*)",
            "ok": ".*"
          }
        ],
        "accepts": [
  
        ],
        "scours": [
          {
            "directory": "/opt/ldm/data",
            "days_old": "3",
            "pattern": "*"
          }
        ],
        "pqacts": [
          {
            "feedtype": "EXP",
            "pattern": "^([0-9]*)_([0-9]*)_stuff_(this|that)\\.gz$",
            "action": "FILE",
            "options": "data/exp/\\1_\\2_stuff\\3.gz",
            "action_args": ""
          }
        ],
        "cronjobs": [
          {
            "name": "newlog",
            "minute": "0",
            "hour": "0",
            "day": "*",
            "month": "*",
            "weekday": "*",
            "user": "ldm",
            "command": "bash -l -c 'ldmadmin newlog' >/opt/ldm/var/logs/cron.ldmadmin.newlog.log 2>&1"
          },
          {
            "name": "scour",
            "minute": "30",
            "hour": "*/4",
            "day": "*",
            "month": "*",
            "weekday": "*",
            "user": "ldm",
            "command": "bash -l -c 'ldmadmin scour' >/opt/ldm/var/logs/cron.ldmadmin.scour.log 2>&1"
          }
        ]
      }
    } 
```

# Recipes

# Author

Scott Macfarlane
Author:: Scott Macfarlane (<scott@gina.alaska.edu>)

and UAF-GINA 


changelog.md
# Changelog

## [0.6.0] 2018-06-04
### added
- added 'auto-start' attribute to control if the ldm process is started and enabled or not.

## [0.5.0] 2018-03-06
### changed
- Default LDM version.
- Contact info.

### added
- Centos 7 support
- gcc compiler

### removed
- build-essentail 

## [0.4.0]
Pre-changelog version.


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/deprecations'
require 'chefspec/berkshelf'
require_relative '../libraries/helpers'

ChefSpec::Coverage.start!

RSpec.configure do |config|
  # Specify the path for Chef Solo to find cookbooks
  # config.cookbook_path = '/var/cookbooks'

  # Specify the path for Chef Solo to find roles
  # config.role_path = '/var/roles'

  # Specify the Chef log_level (default: :warn)
  # config.log_level = :debug

  # Specify the path to a local JSON file with Ohai data
  # config.path = 'ohai.json'

  # Specify the operating platform to mock Ohai data from
  # config.platform = 'ubuntu'

  # Specify the operating version to mock Ohai data from
  # config.version = '12.04'
end
at_exit { ChefSpec::Coverage.report! }


spec/unit/default_spec.rb
require_relative '../spec_helper'

describe 'ldm::default' do
  before do
    allow_any_instance_of(Chef::Recipe).to receive(:queue_exists?).and_return(true)
    Mixlib::ShellOut.stub(:initialize).and_return(shellout)
  end

  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'adds the ldm user' do
    expect(chef_run).to create_user('ldm')
  end

  it 'creates the ldm configuration' do
    expect(chef_run).to create_template('/opt/ldm/etc/ldmd.conf').with(
      owner: 'ldm',
      group: 'ldm',
      mode: 0644,
      variables: { accepts: [], allows: [], requests:[] }
    )
  end

  it 'sets env variables' do
    expect(chef_run).to create_file('/etc/profile.d/ldm.sh').with(
      content: "export PATH=/opt/ldm/bin:$PATH",
      mode: 0644
    )
  end

  it 'creates init scripts' do
    expect(chef_run).to create_template('/etc/init.d/ldm').with(
      source: 'ldm.erb',
      mode: 0755,
      variables: { user: 'ldm', ldmadmin: '/opt/ldm/bin/ldmadmin' }
    )
  end

  it 'starts and enables ldm' do
    expect(chef_run).to enable_service('ldm')
    expect(chef_run).to start_service('ldm')
  end

  context 'queue doesnt exist' do
    before do
      allow_any_instance_of(Chef::Recipe).to receive(:queue_exists?).and_return(false)
    end
    it 'creates the ldm queue' do
      expect(chef_run).to run_execute('create_ldm_queue').with(
        environment: {'PATH' => '/opt/ldm/bin:/usr/bin:$PATH'},
        command: 'ldmadmin mkqueue',
        user: 'ldm',
        group: 'ldm'
      )
    end
  end

  context 'queue exists' do
    before do
      allow_any_instance_of(Chef::Recipe).to receive(:queue_exists?).and_return(true)
    end
    it 'creates the ldm queue' do
      expect(chef_run).to nothing_execute('create_ldm_queue').with(
        environment: {'PATH' => '/opt/ldm/bin:/usr/bin:$PATH'},
        command: 'ldmadmin mkqueue',
        user: 'ldm',
        group: 'ldm'
      )
    end
  end

  context 'source install' do
    before do
      chef_run.node.set['ldm']['install_type'] = 'source'
      chef_run.converge
    end

    it 'includes the source recipe' do
      expect(chef_run).to include_recipe('ldm::_source')
    end
  end

  context 'package install' do
    before do
      chef_run.node.set['ldm']['install_type'] = 'package'
      chef_run.converge
    end

    it 'includes the package recipe' do
      expect(chef_run).to install_package('ldm')
    end
  end
end


test/smoke/default/default_test.rb
# # encoding: utf-8

# Inspec test for recipe imiq-map-app-cookbook::default

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

## test ldm
describe service('ldm') do
  it { should be_installed }
  it { should be_enabled }
  it { should be_running }
end

# test ldm queue
describe file('/opt/ldm/var/queues/ldm.pq') do
  it { should exist }

end


recipes/source.rb

package 'gcc'

directory node['ldm']['install_dir'] do
  user 'ldm'
  group 'ldm'
  recursive true
end

directory "#{node['ldm']['install_dir']}/etc" do
  user 'ldm'
  group 'ldm'
  recursive true
end

remote_file ldm_cache_file do
  source node['ldm']['source']
  checksum node['ldm']['checksum'] if node['ldm']['checksum']
  action :create_if_missing
  owner 'ldm'
  group 'ldm'
  notifies :run, "bash[extract_ldm]", :immediately
end

bash 'extract_ldm' do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  code <<-EOH
  tar xzf #{ldm_cache_file}
  chown -R ldm.ldm #{ldm_src_dir}
  EOH
  not_if { ::File.exists? ldm_src_dir }
  notifies :run, "bash[build_ldm]", :immediately
end

bash 'build_ldm' do
  action :nothing
  user 'ldm'
  group 'ldm'
  cwd ldm_src_dir
  environment node['ldm']['build_env'] if node['ldm']['build_env'].any?
  code <<-EOH
    ./configure --disable-root-actions --disable-static --prefix=#{node['ldm']['install_dir']}
    make install
  EOH
  notifies :run, "bash[run_ldm_root_actions]", :immediately
end


bash 'run_ldm_root_actions' do
  action :nothing
  cwd ldm_src_dir
  code <<-EOH
    make root-actions
  EOH
end


recipes/_source.rb
include_recipe "build-essential"

node['ldm']['build_dependencies'].each do |pkg|
  package pkg
end

directory node['ldm']['install_dir'] do
  user 'ldm'
  group 'ldm'
  recursive true
end

directory "#{node['ldm']['install_dir']}/etc" do
  user 'ldm'
  group 'ldm'
  recursive true
end

remote_file ldm_cache_file do
  source node['ldm']['source']
  checksum node['ldm']['checksum']
  notifies :run, "bash[extract_ldm]"
end

bash 'extract_ldm' do
  action :nothing
  cwd Chef::Config[:file_cache_path]
  code <<-EOH
  tar xzf #{ldm_cache_file}
  chown -R ldm.ldm #{ldm_src_dir}
  EOH
  not_if { ::File.exists? ldm_src_dir }
  notifies :run, "bash[build_ldm]"
end

bash 'build_ldm' do
  action :nothing
  user 'ldm'
  group 'ldm'
  cwd ldm_src_dir
  environment node['ldm']['build_env'] if node['ldm']['build_env'].any?
  code <<-EOH
    ./configure --disable-root-actions --disable-static --prefix=#{node['ldm']['install_dir']}
    make install
  EOH
  notifies :run, "bash[run_ldm_root_actions]"
end

bash 'run_ldm_root_actions' do
  action :nothing
  cwd ldm_src_dir
  code <<-EOH
    make root-actions
  EOH
end


recipes/default.rb
#
# Cookbook Name:: ldm
# Recipe:: default
#
# Copyright (C) 2013 Scott Macfarlane
#
# All rights reserved - Do Not Redistribute
#

node['ldm']['packages'].each do |pkg|
  package pkg do
    action :install
  end
end

include_recipe "ldm::_user"
include_recipe "ldm::#{node['ldm']['install_type']}"

template "#{node['ldm']['install_dir']}/etc/ldmd.conf" do
  source "ldmd.conf.erb"
  owner 'ldm'
  group 'ldm'
  mode 0644
  variables({
    accepts: node['ldm']['accepts'],
    allows: node['ldm']['allows'],
    requests: node['ldm']['requests']
  })
end

template "#{node['ldm']['install_dir']}/etc/scour.conf" do
  source "scour.conf.erb"
  owner 'ldm'
  group 'ldm'
  mode 0644
  variables({
    scours: node['ldm']['scours']
  })
end

template "#{node['ldm']['install_dir']}/etc/pqact.conf" do
  source "pqact.conf.erb"
  owner 'ldm'
  group 'ldm'
  mode 0644
  variables({
    pqacts: node['ldm']['pqacts']
  })
end

file "/etc/profile.d/ldm.sh" do
  content <<-EOL.gsub(/^\W+/, '')
    export PATH=#{node['ldm']['install_dir']}/bin:$PATH
    export LDMHOME=#{node['ldm']['install_dir']}
  EOL
  mode 0644
end

template "/etc/init.d/ldm" do
  source "ldm.erb"
  owner 'root'
  group 'root'
  mode 0755
  variables({
    user: node['ldm']['user'],
    ldmadmin: "#{node['ldm']['install_dir']}/bin/ldmadmin",
    pidfile: "#{node['ldm']['install_dir']}/ldmd.pid",
  })
end

if node['ldm']['auto-start'] then 
  service "ldm" do
    action [ :enable, :start ]
  end
end

# setup some cronjobs
node['ldm']['cronjobs'].each do |cj|
  cron "#{cj['name']}" do
    minute "#{cj['minute']}"
    hour "#{cj['hour']}"
    day "#{cj['day']}"
    month "#{cj['month']}"
    weekday "#{cj['weekday']}"
    user "#{cj['user']}"
    command "#{cj['command']}"
    action :create
  end
end

#Only run this on first installation
execute "create_ldm_queue" do
  environment({"PATH" => "#{node['ldm']['install_dir']}/bin:/usr/bin:$PATH"}) if node['ldm']['install_dir']
  command "#{node['ldm']['install_dir']}/bin/ldmadmin mkqueue"
  user "ldm"
  group "ldm"
  not_if { system("#{node['ldm']['install_dir']}/bin/pqcheck") }
end

file "#{node['ldm']['install_dir']}/var/queues/ldm.pq" do
  mode 0666
  only_if { node['ldm']['global_queue'] }
end
